### PR TITLE
Fix InputBagDynamicReturnTypeExtension

### PR DESCRIPTION
**Pull request by raalderink**
For the InputBag::get, if you add a default string parameter, it will always return a string. This is new in Symfony 5.0.

Fixes issue #91

### CODE REVIEW

File: src/Type/Symfony/InputBagDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+
+final class InputBagDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
```
Comment (ondrejmirtes) [inline src/Type/Symfony/InputBagDynamicReturnTypeExtension.php:14]: The extension needs to be registered in `extension.neon`.

File: composer.json [hunk 2]
```diff
@@ -33,7 +33,7 @@
     "squizlabs/php_codesniffer": "^3.3.2",
     "symfony/console": "^4.0",
     "symfony/framework-bundle": "^4.0",
-    "symfony/http-foundation": "^4.0",
+    "symfony/http-foundation": "^4.0|^5.0",
```
Comment (ondrejmirtes) [inline composer.json:36]: Why this change? Doesn't seem related to the rest.

File: src/Type/Symfony/InputBagDynamicReturnTypeExtension.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		return $methodReflection->getName() === 'get';
+	}
+
+	public function getTypeFromMethodCall(
+		MethodReflection $methodReflection,
+		MethodCall $methodCall,
+		Scope $scope
+	): Type
+	{
+		if (isset($methodCall->args[1]) && $scope->getType($methodCall->args[1]->value) instanceof ConstantStringType) {
```
Comment (ondrejmirtes) [inline src/Type/Symfony/InputBagDynamicReturnTypeExtension.php:32]: I'd rather use this logic:

```php
$argType = $scope->getType($methodCall->args[1]->value);
$isNull = (new NullType())->isSuperTypeOf($argType);
$isString = (new StringType())->isSuperTypeOf($argType);
$compare = $isNull->compareTo($isString);
if ($compare === $isString) {
    return new StringType();
}

return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
```

It will work better in some edge cases:

1) If we have `mixed`, the default return type will be used.
2) If we have `string|null`, the default return type will be used.

File: composer.json (hunk 2 above)
Comment (raalderink) [reply composer.json:36]: The InputBag is new in Symfony 5, didn't exist in 4 yet, so to run the tests we need Symfony 5

Comment (ondrejmirtes) [reply composer.json:36]: There currently isn't a nice way to test it in a 3rd party package, so you can delete the file `tests/Type/Symfony/input_bag_get.php` which isn't very useful currently.

Comment (raalderink) [reply composer.json:36]: Okay, fair, so just delete the tests altogether?

Comment (ondrejmirtes) [reply composer.json:36]: Oh sorry, I haven't noticed there's actually a way here in phpstan-symfony with the ExtensionTestCase :) I just thought you create the file `input_bag_get.php` and let PHPStan analyse it, but right now you're actually checking the return types which is great :) So feel free to add some tests that fail with `mixed`, `string|null` etc. and see how my suggested logic fixes that :)

Comment (raalderink) [reply composer.json:36]: I'm not quite sure what you want me too add there? The `$default` parameter can technically be mixed, however, anything other than a string will throw a deprecation notice. Besides, falling back on default will say `string|null` either way, no matter what you put in `$default`, so I'm not sure if it would matter. I'm fine with adding other types to the test but the result will be the same :P

Comment (ondrejmirtes) [reply composer.json:36]: `mixed` means that it can be a string or null and we can't be sure, so the return type also has to be `string|null`. I just want to test the extension for `mixed` and `string|null` types as the second parameter.

Comment (raalderink) [reply composer.json:36]: PR updated, I think that's what you were looking for?

Comment (ondrejmirtes) [reply composer.json:36]: Approximately :) Thanks.

Comment (ondrejmirtes) [reply composer.json:36]: Please fix the build: https://travis-ci.com/github/phpstan/phpstan-symfony/jobs/372130355

**Comment by ruudk** [comment]
@raalderink Anything I can do to help this PR get merged? I'm having the same issue while upgrading to Symfony 5.

**Comment by ruudk** [comment]
Created a PR on your fork to get rid of the failing test. https://github.com/raalderink/phpstan-symfony/pull/1

**Comment by raalderink** [comment]
@ruudk Thanks, I had lost track of this as I went on holiday

**Comment by ondrejmirtes** [comment]
Hi, the logic in the extension is solid but:

1) I don't get the change in composer.json.
2) The build failed.

**Comment by ruudk** [comment]
@raalderink Are you going to fix this or should I take it over?

**Comment by ruudk** [comment]
@ondrejmirtes @raalderink Let's continue in https://github.com/phpstan/phpstan-symfony/pull/107 which is ✅  and ready to merge 🎉
